### PR TITLE
Switch to use mariabackup instead of just copying files for starter db, fixes #748

### LINKED ISSUE #748: Database corruption on host/docker daemon restart "Failed to start sitename: db service health check timed "
## What happened (or feature request):

We sometimes see "Failed to start sitename: db service health check timed " situations. The log often has something like "InnoDB: The log sequence number 1600607 in ibdata file do not match the log sequence number 10696720565 in the ib_logfiles!" and implies a crash.

Perhaps as in https://stackoverflow.com/questions/19378565/log-sequence-number-in-ibdata-files-does-not-match

This may be happening on a host restart; the db container may not have a reasonable shutdown sequence, so we end up with a corrupted db. We *could* delete the logfiles on start (so they'd be created automatically). There are probably other options. 

**Note: This is almost always caused by a corrupted database, most often in a larger database. Since v0.16.0, this is generally only caused by docker being shut down in an ungraceful way. Unfortunately, both Docker for WIndows and Docker for Mac shut down without notifying the container during upgrade, with a manual Docker exit, or at system shutdown.  To fix this, remove the database with `ddev remove --remove-data` and then `ddev start` and re-import the db. It can be avoided by stopping or remove-ing your projects before letting Docker exit.**

### PR DESCRIPTION
**Pull request by rfay**
## The Problem/Issue/Bug:

OP #748 - Database corruption on docker restart

## How this PR Solves The Problem:

The db container startup script (and base db creation script) used a pretty questionable technique of just saving away a tarball of files. This PR switches to use the normal mariabackup approach, which is much more official.

## Manual Testing Instructions:

* Use the testing technique in https://github.com/drud/ddev/issues/748#issuecomment-400792445

Without this patch, you get corruption after docker hard stop
With this PR a `ddev start` after docker hard stop results in happy db.

## Automated Testing Overview:

I don't know how to create this situation in a test environment predictably, so it's probably best not to try for it. We could force an rm (without the intermediate stop), but that's not the actual described scenario. 

## Related Issue Link(s):
OP #748 

## Release/Deployment notes:

It would be nice if we could tell people "You don't have to stop projects before docker restart or host restart any more." That would be really nice.

Congrats to @andrewfrench for making this whole approach possible, as I was convinced it wasn't something in our realm, thought it was a docker-only problem. Doing the sleuthing to prove this appeared between v0.14.0 and v0.15.0 gave the ammunition to seek a path forward.

### CODE REVIEW

**Comment by rfay** [comment]
I think before pulling I should build an image that will help me test whether the accusation I made against docker (that it doesn't give notice to the container) was even true. It shouldn't be hard and might make me more informed in the future.

**Comment by rfay** [comment]
TLDR; This patch improves mariadb's resiliency and recovery ability after a crash, but the accusation against docker of not giving notice to containers when it shuts down was correct. It just seems that with this change mariadb/mysqld is more likely to recover.

I did some experimentation and it seems that the accusation against docker was not invalid. I tried changing the last lines of docker-entrypoint.sh to the below in order to see the actual behavior of the process

```
echo "Starting mysqld."
tail -f /var/log/mysqld.log &



mysqld &
pid=$!


trap 'echo catching signal at $(adjtimex|grep time.tv) && kill $pid && echo Killed mysqld && sleep 5 && echo sleep done at $(adjtimex|grep time.tv)'  SIGTERM SIGINT

echo "About to wait for mysql pid=$pid"
wait $pid

echo "mysqld pid exited at $(adjtimex|grep time.tv), sleeping" && sleep 5
``` 

With this setup `ddev stop` behaved as you'd expect, 

```
+++ grep time.tv
+++ adjtimex
catching signal at time.tv_sec: 1530578084 time.tv_usec: 932390
Killed mysqld
++ echo catching signal at time.tv_sec: 1530578084 time.tv_usec: 932390
++ kill 6
++ echo Killed mysqld
++ sleep 5
2018-07-03  0:34:44 140651432205032 [Note] mysqld: Normal shutdown
2018-07-03  0:34:44 140651432205032 [Note] Event Scheduler: Purging the queue. 0 events
2018-07-03  0:34:44 140651434662632 [Note] InnoDB: FTS optimize thread exiting.
2018-07-03  0:34:44 140651432205032 [Note] InnoDB: Starting shutdown...
2018-07-03  0:34:45 140651432205032 [Note] InnoDB: Waiting for page_cleaner to finish flushing of buffer pool
2018-07-03  0:34:46 140651432205032 [Note] InnoDB: Shutdown completed; log sequence number 2284808695
2018-07-03  0:34:46 140651432205032 [Note] mysqld: Shutdown complete

+++ adjtimex
+++ grep time.tv
sleep done at time.tv_sec: 1530578089 time.tv_usec: 937627
++ echo sleep done at time.tv_sec: 1530578089 time.tv_usec: 937627
```

However, Just exiting docker resulted in an immediate exit from `ddev logs -s db -f` or from `docker logs -f`; after docker came up again, no graceful shutdown output showed in the logs (before restarting) and then on start we get mysql cleaning up and recovering:

```
Starting mysqld.
About to wait for mysql pid=6
Hope that's ok; if not, decrease some variables in the equation.

Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
Cannot determine thread, fp=0x48400, backtrace may not be correct.
Bogus stack limit or frame pointer, fp=0x48400, stack_bottom=0x7f14565f0000, thread_stack=295936, aborting backtrace.
The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains
information that should help you find out what is causing the crash.
2018-07-03  0:29:53 140652842666888 [Note] mysqld (mysqld 10.1.32-MariaDB) starting as process 6 ...
mysqld: One can only use the --user switch if running as root
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Using mutexes to ref count buffer pool pages
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: The InnoDB memory heap is disabled
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Compressed tables use zlib 1.2.11
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Using SSE crc32 instructions
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Initializing buffer pool, size = 1.0G
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Completed initialization of buffer pool
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: Highest supported file format is Barracuda.
2018-07-03  0:29:53 140652842666888 [Note] InnoDB: The log sequence number 2284808675 in ibdata file do not match the log sequence number 2284808685 in the ib_logfiles!
2018-07-03  0:29:54 140652842666888 [Note] InnoDB: Restoring possible half-written data pages from the doublewrite buffer...
InnoDB: Last MySQL binlog file position 0 206991, file name ./on.000002
2018-07-03  0:29:55 140652842666888 [Note] InnoDB: 128 rollback segment(s) are active.
2018-07-03  0:29:55 140652842666888 [Note] InnoDB: Waiting for purge to start
2018-07-03  0:29:55 140652842666888 [Note] InnoDB:  Percona XtraDB (http://www.percona.com) 5.6.38-83.0 started; log sequence number 2284808685
2018-07-03  0:29:55 140652842666888 [Note] Plugin 'FEEDBACK' is disabled.
2018-07-03  0:29:55 140651434285800 [Note] InnoDB: Dumping buffer pool(s) not yet started
2018-07-03  0:29:55 140652842666888 [Note] Recovering after a crash using on
2018-07-03  0:29:55 140652842666888 [Note] Starting crash recovery...
2018-07-03  0:29:55 140652842666888 [Note] Crash recovery finished.
2018-07-03  0:29:55 140652842666888 [Note] Server socket created on IP: '::'.
2018-07-03  0:29:55 140652842666888 [Warning] 'user' entry 'root@4c129a25133f' ignored in --skip-name-resolve mode.
2018-07-03  0:29:55 140652842666888 [Warning] 'user' entry '@4c129a25133f' ignored in --skip-name-resolve mode.
2018-07-03  0:29:55 140652842666888 [Warning] 'proxies_priv' entry '@% root@4c129a25133f' ignored in --skip-name-resolve mode.
2018-07-03  0:29:55 140652842666888 [Note] mysqld: ready for connections.
Version: '10.1.32-MariaDB'  socket: '/var/tmp/mysql.sock'  port: 3306  MariaDB Server
```


With the large-database test proposed for this issue, a careful review of the startup shows a crash recovery:

```
2018-07-03  2:19:05 140360905730952 [Note] InnoDB: The log sequence number 2284804060 in ibdata file do not match the log sequence number 2284804070 in the ib_logfiles!
2018-07-03  2:19:06 140360905730952 [Note] InnoDB: Restoring possible half-written data pages from the doublewrite buffer...
InnoDB: Last MySQL binlog file position 0 203824, file name ./on.000003
2018-07-03  2:19:07 140360905730952 [Note] InnoDB: 128 rollback segment(s) are active.
2018-07-03  2:19:07 140360905730952 [Note] InnoDB: Waiting for purge to start
2018-07-03  2:19:07 140360905730952 [Note] InnoDB:  Percona XtraDB (http://www.percona.com) 5.6.38-83.0 started; log sequence number 2284804070
2018-07-03  2:19:07 140360905730952 [Note] Plugin 'FEEDBACK' is disabled.
2018-07-03  2:19:07 140359497349864 [Note] InnoDB: Dumping buffer pool(s) not yet started
2018-07-03  2:19:07 140360905730952 [Note] Recovering after a crash using on
2018-07-03  2:19:07 140360905730952 [Note] Starting crash recovery...
2018-07-03  2:19:07 140360905730952 [Note] Crash recovery finished.
```

**Comment by rfay** [comment]
I created the upstream issue https://github.com/docker/for-mac/issues/3047 to point this out to the Docker folks including the repo https://github.com/rfay/containerexittest to demonstrate and explain the problem.

**Comment by rfay** [comment]
@andrewfrench before we pull this could you please do your test and examine the db logs after restarting docker and the `ddev start` following? `ddev logs -s db` - I assume you'll see crash recovery.

**Comment by andrewfrench** [comment]
@rfay, that's correct -- here's a snippet from the db container log after recovery:

```
2018-07-03 15:15:06 140641294121864 [Note] InnoDB: The log sequence number 1617430 in ibdata file do not match the log sequence number 2654764 in the ib_logfiles!
2018-07-03 15:15:06 140641294121864 [Note] InnoDB: Restoring possible half-written data pages from the doublewrite buffer...
InnoDB: Last MySQL binlog file position 0 349785, file name ./on.000001
2018-07-03 15:15:07 140641294121864 [Note] InnoDB: 128 rollback segment(s) are active.
2018-07-03 15:15:07 140641294121864 [Note] InnoDB: Waiting for purge to start
2018-07-03 15:15:07 140641294121864 [Note] InnoDB:  Percona XtraDB (http://www.percona.com) 5.6.38-83.0 started; log sequence number 2654764
2018-07-03 15:15:07 140639885740776 [Note] InnoDB: Dumping buffer pool(s) not yet started
2018-07-03 15:15:07 140641294121864 [Note] Plugin 'FEEDBACK' is disabled.
2018-07-03 15:15:07 140641294121864 [Note] Recovering after a crash using on
2018-07-03 15:15:07 140641294121864 [Note] Starting crash recovery...
2018-07-03 15:15:07 140641294121864 [Note] Crash recovery finished.
```
